### PR TITLE
Add H264RtpDepacketizer

### PR DESCRIPTION
**Pull request by Sean-Der**
Inverse of H264RtpPacketizer. Takes incoming H264 packets and emits H264 NALUs.

### CODE REVIEW

**Comment by Sean-Der** [comment]
@paullouisageneau can/should I add more fields to `Message`? It would be nice to know Duration + Discontinuity

File: src/h264rtpdepacketizer.cpp
Comment (paullouisageneau) [inline src/h264rtpdepacketizer.cpp:130]: I think there are issues with the handling of `messages`. For instance:
- If there is a single timestamp in `rtp_buffer` (for instance a single frame), no new frame is depacketized (because of the break just above), in that case it looks like input messages in `messages` won't be cleared and will leak to the next element in the media processing chain.
- If there are two frames depacketized in a single call, `messages` will be replaced for each frame, so frames will be dropped and only the last one will be passed to the chain.

Comment (paullouisageneau) [reply src/h264rtpdepacketizer.cpp:130]: Maybe I'm missing parts of the logic but if the principle is to flush the current frame when the next timestamp is seen, couldn't such a simple approach do the job for `H264RtpDepacketizer::incoming`?
```cpp
message_vector result;
for (auto message : messages) {
    [...] // check message type and size
    auto p = reinterpret_cast<const RtpHeader *>(message->data());
    if (!rtp_buffer.empty() && current_timestamp != p->timestamp()) {
        result.push_back(buildFrame(rtp_buffer.begin(), rtp_buffer.end()));
        rtp.buffer.clear();
   }
   current_timestamp = p->timestamp();
   rtp_buffer.push_back(std::move(message));
}
messages.swap(result);
```
`current_timestamp` could be a class member (or read from a packet in `rtp_buffer` before the loop).

File: include/rtc/h264rtpdepacketizer.hpp
Comment (paullouisageneau) [inline include/rtc/h264rtpdepacketizer.hpp:33]: The naming convention in libdatachannel for private member variables is camel case with `m` prefix, so I'd actually prefer if it was called `mRtpBuffer` (which allows to unambiguously refer to it without explicit `this->`).

**Review by paullouisageneau — COMMENTED** [review]
> @paullouisageneau can/should I add more fields to `Message`? It would be nice to know Duration + Discontinuity

Yes, feel free to add them, but not directly in `Message` as it adds overhead to every packet in every transport. A `shared_ptr<FrameInfo>` would make sense, like for the reliability information. You should also add the frame timestamp there.

**Comment by paullouisageneau** [comment]
Thank for adding the depacketizer, this is great!

For visibility, this PR partially implements https://github.com/paullouisageneau/libdatachannel/issues/676.

**Comment by walletiger** [comment]
how to use this depacketizer in c_api ?

File: include/rtc/h264rtpdepacketizer.hpp
Comment (Sean-Der) [reply include/rtc/h264rtpdepacketizer.hpp:33]: Renamed

File: src/h264rtpdepacketizer.cpp
Comment (Sean-Der) [reply src/h264rtpdepacketizer.cpp:104]: Fixed

File: src/h264rtpdepacketizer.cpp
Comment (Sean-Der) [reply src/h264rtpdepacketizer.cpp]: Fixed!

File: src/h264rtpdepacketizer.cpp
Comment (Sean-Der) [reply src/h264rtpdepacketizer.cpp]: FIxed!

File: src/h264rtpdepacketizer.cpp
Comment (Sean-Der) [reply src/h264rtpdepacketizer.cpp]: Doesn't `mRtpBuffer` need a copy because it will last across reads?

When I don't copy I get this crash

```
Thread 7 "RTC worker" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7ffff47fa6c0 (LWP 1950035)]
rtc::RtpHeader::timestamp (this=0x0) at /home/sean/workspaces/libdatachannel/src/rtp.cpp:58
58      uint32_t RtpHeader::timestamp() const { return ntohl(_timestamp); }
(gdb) bt
#0  rtc::RtpHeader::timestamp (this=0x0) at /home/sean/workspaces/libdatachannel/src/rtp.cpp:58
#1  0x00007ffff7d38c05 in rtc::H264RtpDepacketizer::incoming(std::vector<std::shared_ptr<rtc::Message>, std::allocator<std::shared_ptr<rtc::Message> > >&, std::function<void (std::shared_ptr<rtc::Message>)> const&) (this=0x476790, messages=std::vector of length 1, capacity 1 = {...}) at /home/sean/workspaces/libdatachannel/src/h264rtpdepacketizer.cpp:116
#2  0x00007ffff7dba38c in rtc::impl::Track::incoming (this=this@entry=0x477540, message=std::shared_ptr<rtc::Message> (empty) = {...})
    at /home/sean/workspaces/libdatachannel/src/impl/track.cpp:146
#3  0x00007ffff7d8da6b in rtc::impl::PeerConnection::dispatchMedia (this=0x4251f0, message=std::shared_ptr<rtc::Message> (use count 6, weak count 0) = {...})
    at /home/sean/workspaces/libdatachannel/src/impl/peerconnection.cpp:599
```

File: src/h264rtpdepacketizer.cpp
Comment (Sean-Der) [reply src/h264rtpdepacketizer.cpp]: Both should be handled now!

> If incoming RTP packets aren't enough to build a frame
`messages.clear()` is called so messages aren't leaked

> two frames in a singe call
I merge the lists now. If a incoming RTP packet results in multiple frames being available it works!

**Comment by Sean-Der** [comment]
@paullouisageneau can I get another review please! Sorry for the delay I will be on top of this now :)

File: src/h264rtpdepacketizer.cpp
Comment (paullouisageneau) [reply src/h264rtpdepacketizer.cpp]: It looks correct now, even if I'm still a bit puzzled by the convoluted approach.

File: src/h264rtpdepacketizer.cpp
Comment (paullouisageneau) [reply src/h264rtpdepacketizer.cpp]: Messages are held as `shared_ptr` in order to keep them until the last reference is removed, so there should be no need to create a copy (also, copying this way loses message metadata like type).

The issue here is that the message is cleared or consumed by a move for some reason, so `size()` is 0 and `data()` ends up returning `nullptr`. `Track` should be the only place were a message is consumed to move it to the user as binary. It might indicate a deeper problem like a message leaking to the user.

File: src/h264rtpdepacketizer.cpp
Comment (Sean-Der) [reply src/h264rtpdepacketizer.cpp:41]: Fixed!

File: src/h264rtpdepacketizer.cpp
Comment (Sean-Der) [reply src/h264rtpdepacketizer.cpp:48]: @paullouisageneau Wasn't aware of it! I just looked a bit and I don't believe they are applicable.

`nalunit.hpp` seems to be just concerned with detecting/splitting NAL units and not the actual understanding of them?

I am all for expanding `nalunit.hpp` to include this logic also though if you want that in this commit.

**Comment by Sean-Der** [comment]
After this commits lands I am going to add `FrameInfo` and a `Opus` depacketizer! After that I can go back and add WHEP support to OBS.

Thanks for merging+reviewing so much @paullouisageneau

File: src/h264rtpdepacketizer.cpp
Comment (paullouisageneau) [reply src/h264rtpdepacketizer.cpp:48]: Good point. No need to expand the logic, but maybe you could only use the header struct from `nalunit.hpp` to read the fields here?

**Review by paullouisageneau — COMMENTED** [review]
@Sean-Der It looks good, thank you for your work! Would you mind replacing the `firstByte` and `secondByte` manipulation with casts to `NalUnitHeader` and `NalUnitFragmentHeader` so it is not implemented twice?

**Comment by Sean-Der** [comment]
@paullouisageneau Done! Can I get another review?

I added one small method to `NalUnitHeader` and was able to drop the firstByte/secondByte.

With some more refactoring/exposing things we could drop even more.

File: src/h264rtpdepacketizer.cpp
Comment (Sean-Der) [reply src/h264rtpdepacketizer.cpp]: @paullouisageneau yea! moved down inside the FU-A check

**Comment by Sean-Der** [comment]
@paullouisageneau Ok I think I got it this time :) Mind taking a look and if this is good going to start Opus + FrameInfo

**Review by paullouisageneau — APPROVED** [review]
It looks good, thank you! If you have the opportunity to add Opus and `FrameInfo` metadata, I would be my pleasure to review it.

**Pull request merged**
